service-workers/redirects/fragment.https.window.js

```diff
@@ -0,0 +1,35 @@
+// META: script=/html/canvas/resources/canvas-tests.js
+// META: script=../service-worker/resources/test-helpers.sub.js
+
+promise_test(async t => {
+  const reg = await service_worker_unregister_and_register(t, "resources/fragment-serviceworker.js", "resources/");
+  // Not add_cleanup as we want this to run after all tests
+  add_completion_callback(async () => await reg.unregister());
+  await wait_for_state(t, reg.installing, 'activated');
+  const frame = await with_iframe("resources/dummy.html");
+  const canvas = frame.contentDocument.body.appendChild(document.createElement("canvas"));
+  canvas.width = 100;
+  canvas.height = 50;
+  const ctx = canvas.getContext("2d");
+
+  // Changing this list requires corresponding changes in resources/fragment-serviceworker.js
```

Comment:
Instead of requiring reasoning across multiple files and keeping things in sync you could have a separate json data file.  That would let you see the entire test input, output, and expectations in a single place.  And things wouldn't get out of sync.  You would just need to have the document and service workers load the json data file using fetch(), etc.